pkg/component/server/konnectivity.go

```diff
@@ -124,10 +154,105 @@ func (k *Konnectivity) writeKonnectivityAgent() error {
 	if err != nil {
 		return fmt.Errorf("failed to write konnectivity agent manifest: %v", err)
 	}
-
 	return nil
 }
 
+func (k *Konnectivity) runLease() {
+	k.done = make(chan bool)
+
+	go func() {
+		logrus.Infof("starting %v lease watcher", serviceName)
+		leaseLock, err := k.newLeaseLock()
+		if err != nil {
+			logrus.Error(err)
+		}
+		k.leaseLock = leaseLock
+		logrus.Debugf("found %v lease holders", k.serverCount)
+
+		ctx := context.Background()
+		defer func() {
+			k.done <- true
+		}()
+		for {
+			k.leaseLock.LeaseRunner(ctx)
+		}
+	}()
+
+	go func() {
+		logrus.Infof("watching %v lease holders", serviceName)
+		ticker := time.NewTicker(10 * time.Second)
+		ctx := context.Background()
+
+		defer func() {
+			k.done <- true
+		}()
+		for {
+			select {
+			case <-ticker.C:
+				observedLeaseHolders := k.leaseLock.CountValidLeaseHolders(ctx)
+				if observedLeaseHolders != k.serverCount {
+					logrus.Debugf("change in %v lease holders detected. refreshing service.", serviceName)
+					k.serverCount = observedLeaseHolders
+
+					// restarting service
+					k.Stop()
+					k.Run()
```

Comment:
As far I can see, `runLease()` will start two go routines that never returns. (both has a `for {}` loop without any return/break or condition to exit the loop). I can not see that `k.Stop()` shuts those go routines down, so they will remain running. The `k.Run()` here will recursively spawn new instances of those go-routines (`Run()` -> `runLease()` -> `Run()` ...).

Do we have a leak here?